### PR TITLE
Update `iam_group` to add check mode support

### PR DESCRIPTION
**Pull request by joshsouza**
##### SUMMARY
This adds check mode support to the `iam_group` module.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
iam_group

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /Users/josh.souza/workspace/infrastructure/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Oct 11 2016, 05:16:02) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

##### ADDITIONAL INFORMATION
Adds check mode support, and restructures code slightly to skip action steps when in check mode.

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/iam_group.py [hunk 1]
```diff
@@ -180,7 +180,7 @@ def compare_attached_group_policies(current_attached_policies, new_attached_poli
     for policy in current_attached_policies:
         current_attached_policies_arn_list.append(policy['PolicyArn'])
 
-    if set(current_attached_policies_arn_list) == set(new_attached_policies):
+    if not set(current_attached_policies_arn_list).symmetric_difference(set(new_attached_policies)):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/iam_group.py]: What is this change for?

File: lib/ansible/modules/cloud/amazon/iam_group.py [hunk 2]
```diff
@@ -282,25 +288,33 @@ def create_or_update_group(connection, module):
...
+                    try:
+                        connection.remove_user_from_group(GroupName=params['GroupName'], UserName=user)
+                    except ClientError as e:
+                        module.fail_json(msg=e.message, exception=traceback.format_exc(),
+                                         **camel_dict_to_snake_dict(e.response))
+                    except ParamValidationError as e:
+                        module.fail_json(msg=e.message, exception=traceback.format_exc())
+        # If there are users to adjust that aren't in the current list, then things have changed
+        # Otherwise the only changes were in purging above
+        if set(users).difference(set(current_group_members_list)):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/iam_group.py]: Why not `if set(users) != set(current_group_members_list)` ?

File: lib/ansible/modules/cloud/amazon/iam_group.py [hunk 3]
```diff
@@ -243,34 +246,37 @@ def create_or_update_group(connection, module):
...
     current_attached_policies = get_attached_policy_list(connection, module, params['GroupName'])
     if not compare_attached_group_policies(current_attached_policies, managed_policies):
+        current_attached_policies_arn_list = []
+        for policy in current_attached_policies:
+            current_attached_policies_arn_list.append(policy['PolicyArn'])
+
         # If managed_policies has a single empty element we want to remove all attached policies
         if purge_policy:
             # Detach policies not present
-            current_attached_policies_arn_list = []
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/iam_group.py:253]: I'm not wholly convinced that the restructuring here is necessary for check mode - the wider changes here make this more difficult to review. Is there a reason that this change is more complicated than putting all the AWS write calls inside an `if module.check_mode` check?

File: lib/ansible/modules/cloud/amazon/iam_group.py (hunk 1 above)
Comment (joshsouza) [reply lib/ansible/modules/cloud/amazon/iam_group.py]: I had a misunderstanding on how set comparisons worked. Will remove

File: lib/ansible/modules/cloud/amazon/iam_group.py (hunk 3 above)
Comment (joshsouza) [reply lib/ansible/modules/cloud/amazon/iam_group.py:253]: The reason for the restructure is that if `purge_policy` was false (basically: don't remove existing policies) we would still flag the group as changed even if no changes were being made, and would still loop through all policies that should be attached and run `attach_group_policy` on it, even if unnecessary. While my change doesn't limit that loop, it does ensure that we only return `changed: True` in cases where action would actually be taken to change things.

I'm ok simply dropping an if statement at the top of the whole block if that's what you'd prefer, but I found this discrepancy while working through testing this module and patched.

File: lib/ansible/modules/cloud/amazon/iam_group.py (hunk 2 above)
Comment (joshsouza) [reply lib/ansible/modules/cloud/amazon/iam_group.py]: The above block of code handles when the `current_group_members_list` contains entries that do not exist in our desired `users` list. Running `set(users).difference()` will result in a list of what users we desire that are not already present. The lower block doesn't need to run at all if you're not adding any users to the group. If we used the simple equality check it'd still run.

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/iam_group.py]: Yeah, sorry, you're right, that's fairly obvious now you point that out. Again, we can use standard set operators for this

```
if set(users) - set(current_group_members_list):
```

**Comment by willthames** [comment]
This looks good to me, but I'd like a second opinion (ideally this module would have tests)

**Review by s-hertel — APPROVED** [review]

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
rebuild_merge

**Comment by joshsouza** [comment]
Thanks folks!

**Pull request merged**